src/pudl/output_diff/cli.py

```diff
@@ -0,0 +1,403 @@
+        con = self.get_engine(fname)
+        return pd.concat(
+            [df for df in pd.read_sql_table(table_name, con, chunksize=100_000)]
+        )
+
+
+class OutputPair:
+    """Represents pair of output.
+
+    Baseline and an experiment.
+    """
+
+    def __init__(
+        self, baseline: OutputBundle, experiment: OutputBundle, args: argparse.Namespace
+    ):
+        """Create new instance with given baseline and experiment."""
+        self.baseline = baseline
+        self.experiment = experiment
+        self.args = args
+
+    def compare_files(self) -> ListDiff:
+        """Produces GenericDiff that contains file basenames."""
+        baseline_files = set(self.baseline.list_files())
+        experiment_files = set(self.experiment.list_files())
+        return ListDiff.from_sets("File", baseline_files, experiment_files)
+
+    def compare_sqlite_databases(self, file_diff: ListDiff) -> list[CategoryDiff]:
+        """Generates series of structural diffs over sqlite databases.
+
+        Returns dictionary where keys are diff categories and values are GenericDiff
+        representing the differences at the given category.
+
+        The following categories of diffs are generated:
+        - filename/table_name, values are table names
+        - filename/table_name/schema, values are (col_name, col_type) tuples
+        - filename/table_name/rowcounts, values are [row_count]
+        - filename/table_name/rows, values are [distinct_row_count]
+        """
+        # TODO(rousik): using GenericDiff for rowcounts and rows are not ideal,
+        # perhaps better approach would be to have type-specific diffs that allow
+        # either numbers for left/right (for rowcount) or panda dataframes for rows.
+        # But let's work with the hacky thing for now.
+        output_diffs = []
+
+        def maybe_append_diff(diff: CategoryDiff):
+            """Appends diff to output_diffs if there are actual differences."""
+            if not diff.sides_equal():
+                output_diffs.append(diff)
+                diff.print_diff()
+
+        for fname in file_diff.both:
+            logger.info(f"Analyzing file {fname}")
+            if not fname.endswith(".sqlite"):
+                # This file is not supported by this comparison function
+                continue
+
+            table_diff = self.compare_tables(fname)
+            maybe_append_diff(table_diff)
+            for table_name in table_diff.both:
+                schema_diff = self.compare_table_schemas(fname, table_name)
+                maybe_append_diff(schema_diff)
+                # We could consider comparing tables that have divergent schemas,
+                # but that would be trickier.
+                if schema_diff.sides_equal():
+                    rc_diff = self.compare_table_rowcounts(fname, table_name)
+                    maybe_append_diff(rc_diff)
+                    # if rc_diff.sides_equal():
```

Comment:
It might be worth diffing rows even if the row counts are different - I could imagine a situation where we're both missing data in the experiment and also have rows with completely different data.

Reply:
I've decided to drop row-level comparisons entirely for now. I'm quite aware that this is going to be extra complicated and we might need to come up with some system to compare these without blowing up the memory for extremely large tables, but I think this is beyond the scope of the initial prototype. Perhaps we can compare row-counts and schemas and assume that most changes should impact those high-level cheap metrics.

Loading both tables into dataframes and comparing those directly is likely not feasible for very large tables with lots of rows, so I'm thinking that we could employ some kind of either statistical approach (sample bunch of records from either side and see if they are present on the other side), or full hashing approach where we could scan the two tables and calculate hash for each row (hash of hashes of columns) and compare those, or even do the above with sharded approach where we take subset of rows based on some hash-key. This might be necessary to support very large tables, but right now I think it's a bit too early to optimize in this space. Let's see how this goes first before embarking on algorithmically complex solutions.